proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+specified.
+
+```swift
+let package = Package(
+    name: "Example",
+    traits: [
+        "Foo",
+        Trait(
+            name: "Bar",
+            enabledTraits: [ // Other traits that are enabled when this trait is being enabled
+                "Foo"
+            ]
+        )
+    ],
+    defaultTraits: [ // The set of default enabled traits
+        "Foo",
+    ],
+    /// ...
+)
+```
+
+When depending on a package a set of enabled traits can be passed and whether the
+default set of traits are disabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: ["SomeTrait"],
+            disableDefaultTraits: true
+        )
+    ),
+]
+```
+
+Another common scenario is to enable a trait of a dependency only when a trait
+of the package is enabled. The below example enables the `SomeOtherTrait` when
+the `Foo` trait of this package is enabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: [
+                "SomeTrait",
+                EnabledTrait("SomeOtherTrait", condition: .when(traits: ["Foo"])),
+            ]
+        )
+    ),
+]
+```
+
+Conditional dependencies are specified per target and extend the current
+`condition` syntax which is used for specifying platform dependent dependencies.
+
+```swift
+targets: [
+    .target(
+        name: "SomeTarget",
+        dependencies: [
+            .product(
+                name: "SomeProduct",
+                package: "SomePackage",
+                condition: .when(traits: ["Foo"])
+            ),
+        ]
+    )
+]
+```
+
+Lastly, code can be conditionally compiled by checking if a trait is enabled.
+This can be used for both optional dependencies by surrounding the `import`
+statements in a trait check and for regular code where you want to modify its
+behaviour depending on the enabled traits.
+
+```swift
+#if TRAIT_Foo
+import SomeDependency
+#endif
+
+func hello() {
+    #if TRAIT_Foo
+    Foo.hello()
+    #else
+    print("Hello")
+    #endif
+}
+```
+
+At this point, it is important to talk about the trait unification across the
+entire dependency graph. After dependency resolution the union of enabled traits
+per package is calculated. This is then used to determine both the enabled
+optional dependencies and the enabled traits for the compile time checks. Since
+the enabled traits of a dependency are specified on a per package level and not
+from the root of the tree, any combination of enabled traits must be supported.
+A consequence of this is that all traits **must** be additive. Enabling a trait
+**must never** disable functionality i.e. remove API or lead to any other
+SemVer-incompatible change.
```

Comment:
Is this enforced somehow?

IMO it seems like the user experience of developing a library which offers traits should be more like availability guards than compiler guards.

Reply:
In theory we could use the API breaking change checker to check the APIs between various traits to assure nothing removes API. IMO, compiler guards are exactly what package traits are but I would love if compiler guards could be spelled nicer in the language more akin to what Rust does i.e. only at the declaration and not surrounding the whole body.

Reply:
I wouldn't trust the API breakage check to check any such thing; the false positive rate of API breakage reports remains appalling (yes, bugs have been filed 🙂).